backends/platform/android/touchcontrols.cpp

```diff
@@ -0,0 +1,362 @@
+		return;
+	}
+
+	if (dY > abs(dX)) {
+		state.main = Common::KEYCODE_DOWN;
+		state.clip = Common::Rect(256, 0, 384, 128);
+	} else if (dX > abs(dY)) {
+		state.main = Common::KEYCODE_RIGHT;
+		state.clip = Common::Rect(128, 0, 256, 128);
+	} else if (-dY > abs(dX)) {
+		state.main = Common::KEYCODE_UP;
+		state.clip = Common::Rect(0, 0, 128, 128);
+	} else if (-dX > abs(dY)) {
+		state.main = Common::KEYCODE_LEFT;
+		state.clip = Common::Rect(384, 0, 512, 128);
+	} else {
+		return;
+	}
+
+	if (sqNorm > 20000) {
+		state.modifier = Common::KEYCODE_LSHIFT;
+	}
+
+}
+
+void TouchControls::touchToCenterState(int dX, int dY, FunctionState &state) {
+	int sqNorm = dX * dX + dY * dY;
+	if (sqNorm < 50 * 50) {
+		state.main = Common::KEYCODE_RETURN;
+	}
+}
+
+void TouchControls::touchToRightState(int dX, int dY, FunctionState &state) {
+	if (dX * dX + dY * dY < 100 * 100)
+		return;
+
+	if (dY > abs(dX)) {
+		// down
+		state.main = Common::KEYCODE_PAGEDOWN;
+		state.clip = Common::Rect(256, 0, 384, 128);
+		return;
+	} else if (-dY > abs(dX)) {
+		// up
+		state.main = Common::KEYCODE_PAGEUP;
+		state.clip = Common::Rect(0, 0, 128, 128);
+		return;
+	}
+
+	static Common::KeyCode keycodes[5] = {
+		Common::KEYCODE_i, Common::KEYCODE_p, // left zone
+		Common::KEYCODE_INVALID, // center
+		Common::KEYCODE_u, Common::KEYCODE_l // right zone
+	};
+
+	static int16 clips[5][4] = {
+		{   0, 128, 128, 256 }, // i
+		{ 128, 128, 256, 256 }, // p
+		{   0,   0,   0,   0 }, // center
+		{ 256, 128, 384, 256 }, // u
+		{ 384, 128, 512, 256 }  // l
+	};
+	static const unsigned int offset = (ARRAYSIZE(keycodes) - 1) / 2;
+
+	int idx = (dX / 100) + offset;
+	if (idx < 0) {
+		idx = 0;
+	}
+	if (idx >= ARRAYSIZE(keycodes)) {
+		idx = ARRAYSIZE(keycodes) - 1;
+	}
+	state.main = keycodes[idx];
+	state.clip = Common::Rect(clips[idx][0], clips[idx][1], clips[idx][2], clips[idx][3]);
+}
+
+TouchControls::FunctionBehavior TouchControls::functionBehaviors[TouchControls::kFunctionMax+1] = {
+	{ touchToJoystickState, false, .2f, .5f },
+	{ touchToCenterState,   true,  .5f, .5f },
+	{ touchToRightState,    true,  .8f, .5f }
+};
+
+static GLES8888Texture *loadBuiltinTexture(const char *filename) {
+	Common::ArchiveMemberPtr member = SearchMan.getMember(filename);
+	Common::SeekableReadStream *str = member->createReadStream();
+	Image::TGADecoder dec;
+	dec.loadStream(*str);
+	const void *pixels = dec.getSurface()->getPixels();
+
+	GLES8888Texture *ret = new GLES8888Texture();
+	uint16 w = dec.getSurface()->w;
+	uint16 h = dec.getSurface()->h;
+	uint16 pitch = dec.getSurface()->pitch;
+	ret->allocBuffer(w, h);
+	ret->updateBuffer(0, 0, w, h, pixels, pitch);
+
+	delete str;
+	return ret;
+}
+
+void TouchControls::init(int width, int height) {
+	_arrows_texture = loadBuiltinTexture("arrows.tga");
+	_screen_width = width;
+	_screen_height = height;
+}
+
+TouchControls::Pointer *TouchControls::getPointerFromId(int ptrId, bool createNotFound) {
+	unsigned int freeEntry = -1;
```

Comment:
I'd recommend using `uint` here and below for brevity

Reply:
Done.